src/feat/feature-window.h

```diff
@@ -0,0 +1,207 @@
+// feat/feature-window.h
+
+// Copyright 2009-2011  Karel Vesely;  Petr Motlicek;  Saarland University
+//           2014-2016  Johns Hopkins University (author: Daniel Povey)
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_FEAT_FEATURE_WINDOW_H_
+#define KALDI_FEAT_FEATURE_WINDOW_H_
+
+#include <map>
+#include <string>
+
+#include "matrix/matrix-lib.h"
+#include "util/common-utils.h"
+#include "base/kaldi-error.h"
+
+namespace kaldi {
+/// @addtogroup  feat FeatureExtraction
+/// @{
+
+struct FrameExtractionOptions {
+  BaseFloat samp_freq;
+  BaseFloat frame_shift_ms;  // in milliseconds.
+  BaseFloat frame_length_ms;  // in milliseconds.
+  BaseFloat dither;  // Amount of dithering, 0.0 means no dither.
+  BaseFloat preemph_coeff;  // Preemphasis coefficient.
+  bool remove_dc_offset;  // Subtract mean of wave before FFT.
+  std::string window_type;  // e.g. Hamming window
+  bool round_to_power_of_two;
+  bool snip_edges;
+  // Maybe "hamming", "rectangular", "povey", "hanning"
+  // "povey" is a window I made to be similar to Hamming but to go to zero at the
+  // edges, it's pow((0.5 - 0.5*cos(n/N*2*pi)), 0.85)
+  // I just don't think the Hamming window makes sense as a windowing function.
```

Comment:
The Hamming window may be optimal in that very particular sense if you
choose from the class of raised cosines plus a constant offset.  But in
this application there is no constraint to that functional form.  It's like
saying "my knife is an optimal shape, given that it has to have a
rectangular cross section".   The Hamming window has discontinuities at the
edges, which means that a low-frequency signal when windowed like that will
generate very high-frequency components, and vice versa.  So for me this is
a case, of I'm not going jump off a bridge just because everyone else is
doing so.


On Mon, Apr 11, 2016 at 12:13 AM, Kirill Katsnelson <
notifications@github.com> wrote:

> In src/feat/feature-window.h
> <https://github.com/kaldi-asr/kaldi/pull/679#discussion_r59161940>:
>
> > +/// @{
> > +
> > +struct FrameExtractionOptions {
> > +  BaseFloat samp_freq;
> > +  BaseFloat frame_shift_ms;  // in milliseconds.
> > +  BaseFloat frame_length_ms;  // in milliseconds.
> > +  BaseFloat dither;  // Amount of dithering, 0.0 means no dither.
> > +  BaseFloat preemph_coeff;  // Preemphasis coefficient.
> > +  bool remove_dc_offset;  // Subtract mean of wave before FFT.
> > +  std::string window_type;  // e.g. Hamming window
> > +  bool round_to_power_of_two;
> > +  bool snip_edges;
> > +  // Maybe "hamming", "rectangular", "povey", "hanning"
> > +  // "povey" is a window I made to be similar to Hamming but to go to zero at the
> > +  // edges, it's pow((0.5 - 0.5*cos(n/N*2*pi)), 0.85)
> > +  // I just don't think the Hamming window makes sense as a windowing function.
>
> Just wondering, why? It makes certain sense in the frequency domain,
> minimizing the closest side-lobe of the signal, although it does not have a
> great roll-off, so the second side-lobe contributes significantly. As a
> guess, it might provide some win with high-resolution bins--but in any
> case, windowing is a matter of engineering rather than science, and I am
> just interested in your opinion here. “Makes no sense” is a strong
> statement, indeed!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/679/files/8e7c759093ca984eeab029d0c6c1393249f1a158#r59161940>
>